iceoryx_hoofs/source/posix_wrapper/named_semaphore.cpp

```diff
@@ -0,0 +1,221 @@
+        case EMFILE:
+            LogError() << "The per-process limit of file descriptor exceeded while creating the semaphore \"" << name
+                       << "\"";
+            return cxx::error<SemaphoreError>(SemaphoreError::FILE_DESCRIPTOR_LIMIT_REACHED);
+        case ENFILE:
+            LogError() << "The system wide limit of file descriptor exceeded while creating the semaphore \"" << name
+                       << "\"";
+            return cxx::error<SemaphoreError>(SemaphoreError::FILE_DESCRIPTOR_LIMIT_REACHED);
+        case ENOMEM:
+            LogError() << "Insufficient memory to create the semaphore \"" << name << "\".";
+            return cxx::error<SemaphoreError>(SemaphoreError::OUT_OF_MEMORY);
+        default:
+            LogError() << "This should never happen. An unknown error occurred while creating the semaphore \"" << name
+                       << "\".";
+            return cxx::error<SemaphoreError>(SemaphoreError::UNDEFINED);
+        }
+    }
+
+    constexpr bool HAS_OWNERSHIP = true;
+    uninitializedSemaphore.emplace(result.value().value, name, HAS_OWNERSHIP);
+
+    return cxx::success<>();
+}
+
+cxx::expected<SemaphoreError>
+NamedSemaphoreBuilder::create(cxx::optional<NamedSemaphore>& uninitializedSemaphore) noexcept
+{
+    if (!cxx::isValidFileName(m_name))
+    {
+        LogError() << "The name \"" << m_name << "\" is not a valid semaphore name.";
+        return cxx::error<SemaphoreError>(SemaphoreError::INVALID_NAME);
+    }
+
+    if (m_initialValue > IOX_SEM_VALUE_MAX)
+    {
+        LogError() << "The semaphores \"" << m_name << "\" initial value of " << m_initialValue
+                   << " exceeds the maximum semaphore value " << IOX_SEM_VALUE_MAX;
+        return cxx::error<SemaphoreError>(SemaphoreError::SEMAPHORE_OVERFLOW);
+    }
+
+    if (m_openMode == OpenMode::OPEN_EXISTING)
+    {
+        auto result = IOX_TRY(tryOpenExistingSemaphore(uninitializedSemaphore, m_name), result);
+
+        if (!result.value())
+        {
+            LogError() << "Unable to open semaphore since no semaphore with the name \"" << m_name << "\" exists.";
+            return cxx::error<SemaphoreError>(SemaphoreError::NO_SEMAPHORE_WITH_THAT_NAME_EXISTS);
+        }
+        return cxx::success<>();
+    }
+    else if (m_openMode == OpenMode::OPEN_OR_CREATE)
+    {
+        auto result = IOX_TRY(tryOpenExistingSemaphore(uninitializedSemaphore, m_name), result);
+
+        if (result.value())
+        {
+            return cxx::success<>();
+        }
+
+        return createSemaphore(uninitializedSemaphore, m_name, m_openMode, m_permissions, m_initialValue);
+    }
+    else if (m_openMode == OpenMode::EXCLUSIVE_CREATE)
+    {
+        return createSemaphore(uninitializedSemaphore, m_name, m_openMode, m_permissions, m_initialValue);
+    }
+
+    // if OpenMode::PURGE_AND_CREATE, written outside of if statement to avoid no return value warning
```

Comment:
This could also be done by returning an error at the end but then this would basically be dead code.

Not sure what would be the best solution here since now it would be possible to add a new value to `OpenMode` and the compiler would not complain and handle it the same as `OpenMode::PURGE_AND_CREATE`. Most probably this would be detected in a test.

I think ideally we would have something like https://en.cppreference.com/w/cpp/utility/unreachable (please don't start implementing it; although it can be done with C++11, this is just me thinking loud ;)) and use a plain switch statement.